pkgs/development/ocaml-modules/zmq/default.nix

```diff
@@ -1,9 +1,16 @@
-{ stdenv, fetchFromGitHub, buildDunePackage, czmq, stdint }:
+{ stdenv, fetchFromGitHub, ocaml, findlib, dune, czmq, stdint }:
 
-buildDunePackage rec {
-  pname = "zmq";
-  version = "20180726";
+if !stdenv.lib.versionAtLeast ocaml.version "4.03"
+then throw "zmq is not available for OCaml ${ocaml.version}"
+else
+
+let __dune = dune; in
+let dune = __dune.override { ocamlPackages = { inherit ocaml findlib; }; };
```

Comment:
So this is the fix, is it? Can you add a comment explaining how this is necessary that dune is built with the exact same version of OCaml and findlib? In the past, you told me that `dune` was not part of `ocamlPackages` because it was an independent tool but in the end, if it had been part of `ocamlPackages` none of this fix would have been necessary.

Reply:
Also, there is no reason why the compiler used to build the build system would need to be the same as the compiler used to build the software.

Maybe we should think more about cross-compilation.

Reply:
Both reasons are invalid:

 1. I do it here because the builder does not cover this case yet: we first need to build the build-system, and then run it to build the package. Since in the current state, both must be done in the same environment, there is only one version of OCaml available.

 2. Some people do use `opam` to do cross-compilation. In particular they manage to run `ppx` at build-time.